### PR TITLE
Implemented FileManager.delegate and cross-device move

### PR DESCRIPTION
**Pull request by amosavian**
- Added `FileManager.delegate` functionality
- Updated `FileManagerDelegate` default implementation method signature to match protocol methods.
- Fixed TODO: Implemented cross-device move in `moveItem()`
- Fixed TODO: Error handling if fts_read fails in `removeItem()`

### CODE REVIEW

**Comment by amosavian** [comment]
@parkera would you mind reviewing this PR?

File: Foundation/FileManager.swift [hunk 1]
```diff
@@ -564,104 +562,268 @@ open class FileManager : NSObject {
...
         guard !self.fileExists(atPath: dstPath) else {
             throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.fileWriteFileExists.rawValue, userInfo: [NSFilePathErrorKey : NSString(dstPath)])
         }
         try _fileSystemRepresentation(withPath: srcPath, andPath: dstPath, {
             if rename($0, $1) != 0 {
                 if errno == EXDEV {
-                    // TODO: Copy and delete.
-                    NSUnimplemented("Cross-device moves not yet implemented")
+                    try _copyOrLinkDirectoryHelper(atPath: srcPath, toPath: dstPath) { (srcPath, dstPath, fileType) in
+                        guard shouldMoveItemAtPath(srcPath, toPath: dstPath, isURL: isURL) else {
```
[the remark this answers is not in the document]
Comment (amosavian) [reply Foundation/FileManager.swift]: No, for two reason:
1. This body is called only when `rename()` fails
2. This is called for every file for recursive copy/move

[the remark this answers is not in the document]
Comment (kperryua) [reply Foundation/FileManager.swift]: FWIW, Darwin's implementation doesn't ask "should move" for every subitem when it falls back to a copy. The delegate-calling semantics of the operation is identical in either case. As such, the initial `shouldMove` check should be sufficient confirmation for either the `rename` or the recursive copy.

File: Foundation/FileManager.swift [hunk 2]
```diff
@@ -564,104 +562,268 @@ open class FileManager : NSObject {
...
                     fts_close(stream)
                 }
 
-                var current = fts_read(stream)
-                while current != nil {
-                    switch Int32(current!.pointee.fts_info) {
+                while let current = fts_read(stream)?.pointee {
+                    
+                    
+                    let itemPath = NSString(bytes: current.fts_path,
```
Comment (amosavian) [inline Foundation/FileManager.swift]: @spevans Can I replace this with `String.init(cString:)`?

Comment (spevans) [reply Foundation/FileManager.swift]: Best to use `self.string(withFileSystemRepresentation:length:)` to convert to a `String`.

**Comment by amosavian** [comment]
Changes:

- Using `string(withFileSystemRepresentation:length:)` when possible
- `shouldMoveItemAtPath` will be called only once for recursive operation
- Adding `"NSSourceFilePathErrorKey"`, `"NSDestinationFilePath"` and `"NSUserStringVariant"` keys to copy/move errors to match Darwin version

cc @kperryua @spevans

**Comment by amosavian** [comment]
Oh, I removed tests from commit accidentally :(

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

**Comment by amosavian** [comment]
Hi, is it normal to have a pull request open without and feedback/activity? @parkera

**Comment by kperryua** [comment]
Apologies, that's my fault. I dropped the ball on following up after my previous feedback. I'll take another look at this today.

**Comment by amosavian** [comment]
@kperryua pardon me for disturbing. But it’s hard to keep unmerged PRs synchronized.

**Comment by amosavian** [comment]
@parkera pardon me for disturbing you repeatedly. Would you mind review this PR?

**Comment by parkera** [comment]
If @kperryua approves that's as good as my approval.

**Comment by amosavian** [comment]
They approved as I see.

**Comment by spevans** [comment]
@swift-ci test and merge

**Pull request merged**